File: Windows/BotSpeaker/RecallRunManager.cs
using System.Text.Json.Nodes;

namespace BotSpeaker;

/// <summary>App-owned CLI meeting plans; navigation and CLI exit do not interrupt a run.</summary>
public sealed class RecallRunManager(Func<string, JsonObject, Task<JsonObject>> request)
{
    private sealed class Run(RecallMeetingTurn[] turns)
    {
        public readonly RecallMeetingTurn[] Turns = turns;
        public readonly string Id = Guid.NewGuid().ToString();
        public string Status = "created";
        public string? Error;
        public int Index = -1;
        public bool Skip;
        public CancellationTokenSource? Cancellation;
        public Task? Task;
        public JsonObject Snapshot() => new() { ["id"] = Id, ["status"] = Status,
            ["turnIndex"] = Index, ["turnCount"] = Turns.Length, ["error"] = Error };
    }
    private readonly Dictionary<string, Run> runs = [];
    public bool HasActiveRuns => runs.Values.Any(run => run.Task is { IsCompleted: false });
    public JsonArray Snapshots() => new(runs.Values.Select(run => (JsonNode)run.Snapshot()).ToArray());
    public async Task StopAllAsync()
    {
        foreach (var run in runs.Values) run.Cancellation?.Cancel();
        await Task.WhenAll(runs.Values.Select(run => run.Task ?? Task.CompletedTask));
    }
    public async Task StopForBotsAsync(HashSet<string> bots)
    {
        var affected = runs.Values.Where(run => run.Turns.Any(turn => bots.Contains(turn.BotId))).ToArray();
        foreach (var run in affected) run.Cancellation?.Cancel();
        await Task.WhenAll(affected.Select(run => run.Task ?? Task.CompletedTask));
    }
    public async Task<JsonObject> HandleAsync(string action, JsonObject body)
    {
        if (action == "meeting-create")
        {
            var items = body["turns"] as JsonArray ?? throw new AppException("Provide a JSON plan with a turns array.");
            if (items.Count is < 1 or > 450) throw new AppException("Provide between 1 and 450 turns.");
            var turns = items.Select(item => {
                if (!Guid.TryParse(item?["botId"]?.GetValue<string>(), out var bot)) throw new AppException("Every turn needs a bot UUID.");
                var voice = item?["voice"]?.GetValue<string>()?.Trim();
                var text = item?["text"]?.GetValue<string>()?.Trim();
                if (string.IsNullOrEmpty(voice) || string.IsNullOrEmpty(text)) throw new AppException("Every turn needs voice and text.");
                return new RecallMeetingTurn(bot.ToString(), voice, text);
            }).ToArray();
            var created = new Run(turns);
            runs.Add(created.Id, created);
            return new() { ["ok"] = true, ["meeting"] = created.Snapshot() };
        }
        var id = body["id"]?.GetValue<string>() ?? "";
        if (!runs.TryGetValue(id, out var current)) throw new AppException("Unknown Recall meeting plan.");
        switch (action)
        {
            case "meeting-status": break;
            case "meeting-start":
                if (current.Task is { IsCompleted: false } || current.Status == "starting") throw new AppException("Meeting is already running.");
                if (runs.Values.Any(run => run.Task is { IsCompleted: false } &&
                    run.Turns.Any(turn => current.Turns.Any(other => other.BotId == turn.BotId))))
                    throw new AppException("A speaker bot is already used by another running meeting plan.");
                current.Status = "starting";
                current.Cancellation = new();
                current.Error = null; current.Index = -1; current.Skip = false;
                current.Task = ExecuteAsync(current);
                break;
            case "meeting-stop":
                current.Cancellation?.Cancel();
                if (current.Task != null) await current.Task;
                else current.Status = "stopped";
                break;
            case "meeting-skip":
                if (current.Status != "running") throw new AppException("A turn must be running to skip it.");
                current.Skip = true;
                break;
            default: throw new AppException("Unknown Recall meeting action.");
        }
        return new() { ["ok"] = true, ["meeting"] = current.Snapshot() };
    }
    private async Task ExecuteAsync(Run run)
    {
        try
        {
            var token = run.Cancellation!.Token;
            var response = await request("list", new()).WaitAsync(token);
            token.ThrowIfCancellationRequested();
            var ready = response["bots"]!.AsArray().Where(bot => bot?["status"]?.GetValue<string>() == "in_call_recording")
                .Select(bot => bot!["id"]!.GetValue<string>()).ToHashSet();
            if (run.Turns.Any(turn => !ready.Contains(turn.BotId))) throw new AppException("Every speaker bot must be in_call_recording before starting.");
            run.Status = "preparing";
            await RecallTurnRunner.RunAsync(run.Turns, request, index => {
                run.Index = index; run.Skip = false; run.Status = "running";
            }, run.Cancellation!.Token, () => run.Skip, index => run.Index = index);
            run.Status = "finished";
        }
        catch (OperationCanceledException) { run.Status = "stopped"; }
        catch (Exception error) { run.Status = "failed"; run.Error = error.Message; }
        finally { run.Cancellation?.Dispose(); run.Cancellation = null; }
    }
}

File: macOS/BotSpeaker/RecallRunManager.swift
import Foundation
import Observation

struct RecallMeetingTurn {
    let botID: String
    let voice: String
    let text: String
}

typealias RecallRequest = @MainActor (String, [String: Any]) async throws -> [String: Any]

/// Shared by GUI and CLI: prepare the whole script before dispatching any audio.
@MainActor
enum RecallTurnRunner {
    static func run(_ turns: [RecallMeetingTurn], request: RecallRequest,
                    progress: (Int) -> Void, skipRequested: () -> Bool = { false },
                    preparing: (Int) -> Void = { _ in },
                    completed: (Int, [String: Any], Bool) -> Void = { _, _, _ in }) async throws {
        var pending: [String] = []
        do {
            for (index, turn) in turns.enumerated() {
                try Task.checkCancellation()
                preparing(index)
                let response = try await request("prepare", ["botId": turn.botID, "voice": turn.voice, "text": turn.text])
                guard let id = (response["job"] as? [String: Any])?["id"] as? String else { throw AppError("Recall did not return a speech job.") }
                pending.append(id)
                try await wait(id, expected: "prepared", request: request)
            }
            for (index, id) in Array(pending).enumerated() {
                try Task.checkCancellation()
                progress(index)
                _ = try await request("dispatch", ["id": id])
                try await wait(id, expected: "finished_dispatching", request: request, skipRequested: skipRequested)
                let response = try await request("status", [:])
                let job = (response["jobs"] as? [[String: Any]])?.first { $0["id"] as? String == id } ?? [:]
                completed(index, job, skipRequested())
                pending.removeAll { $0 == id }
            }
        } catch {
            // This cleanup is local cancellation, so it also works in a cancelled task.
            for id in pending { _ = try? await request("cancel", ["id": id]) }
            throw error
        }
    }

    private static func wait(_ id: String, expected: String, request: RecallRequest,
                             skipRequested: () -> Bool = { false }) async throws {
        while true {
            try Task.checkCancellation()
            if skipRequested() { _ = try await request("cancel", ["id": id]); return }
            let response = try await request("status", [:])
            guard let job = (response["jobs"] as? [[String: Any]])?.first(where: { $0["id"] as? String == id }) else { throw AppError("Speech job was lost.") }
            let status = job["status"] as? String ?? ""
            if status == expected { return }
            if ["failed", "cancelled"].contains(status) { throw AppError(job["error"] as? String ?? "Speech was cancelled.") }
            try await Task.sleep(for: .milliseconds(20))
        }
    }
}

@MainActor @Observable
final class RecallRunManager {
    @MainActor @Observable
    fileprivate final class Run {
        let id = UUID().uuidString
        let turns: [RecallMeetingTurn]
        var status = "created"
        var error: String?
        var index = -1
        var skip = false
        @ObservationIgnored var task: Task<Void, Never>?
        init(_ turns: [RecallMeetingTurn]) { self.turns = turns }
        var snapshot: [String: Any] {
            ["id": id, "status": status, "turnIndex": index, "turnCount": turns.count, "error": error as Any? ?? NSNull()]
        }
    }
    private var runs: [Run] = []
    var snapshots: [[String: Any]] { runs.map(\.snapshot) }
    var hasActiveRuns: Bool { runs.contains { ["starting", "preparing", "running"].contains($0.status) } }

    func stopAll() async {
        let tasks = runs.compactMap(\.task)
        for task in tasks { task.cancel() }
        for task in tasks { await task.value }
    }
    func stopForBots(_ ids: Set<String>) async {
        let tasks = runs.filter { $0.turns.contains { ids.contains($0.botID) } }.compactMap(\.task)
        for task in tasks { task.cancel() }
        for task in tasks { await task.value }
    }
    func handle(_ action: String, _ body: [String: Any], request: @escaping RecallRequest) async throws -> [String: Any] {
        if action == "meeting-create" {
            guard let items = body["turns"] as? [[String: Any]], (1...450).contains(items.count) else { throw AppError("Provide between 1 and 450 turns in a turns array.") }
            let turns = try items.map { item -> RecallMeetingTurn in
                guard let id = item["botId"] as? String, let bot = UUID(uuidString: id) else { throw AppError("Every turn needs a bot UUID.") }
                let voice = (item["voice"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                let text = (item["text"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                guard !voice.isEmpty, !text.isEmpty else { throw AppError("Every turn needs voice and text.") }
                return RecallMeetingTurn(botID: bot.uuidString.lowercased(), voice: voice, text: text)
            }
            let run = Run(turns); runs.append(run)
            return ["ok": true, "meeting": run.snapshot]
        }
        guard let id = body["id"] as? String, let run = runs.first(where: { $0.id == id }) else { throw AppError("Unknown Recall meeting plan.") }
        switch action {
        case "meeting-status": break
        case "meeting-start":
            guard run.task == nil else { throw AppError("Meeting is already running.") }
            let bots = Set(run.turns.map(\.botID))
            guard !runs.contains(where: { $0.task != nil && $0.turns.contains { bots.contains($0.botID) } }) else { throw AppError("A speaker bot is already used by another running meeting plan.") }
            run.status = "starting"; run.error = nil; run.index = -1; run.skip = false
            run.task = Task {
                defer { run.task = nil }
                do {
                    let response = try await request("list", [:])
                    try Task.checkCancellation()
                    let ready = Set((response["bots"] as? [[String: Any]] ?? []).filter { $0["status"] as? String == "in_call_recording" }.compactMap { $0["id"] as? String })
                    guard bots.isSubset(of: ready) else { throw AppError("Every speaker bot must be in_call_recording before starting.") }
                    run.status = "preparing"
                    try await RecallTurnRunner.run(run.turns, request: request, progress: { index in
                        run.index = index; run.skip = false; run.status = "running"
                    }, skipRequested: { run.skip }, preparing: { run.index = $0 })
                    run.status = "finished"
                } catch {
                    run.status = Task.isCancelled ? "stopped" : "failed"
                    if !Task.isCancelled { run.error = error.localizedDescription }
                }
            }
        case "meeting-stop":
            if let task = run.task { task.cancel(); await task.value } else { run.status = "stopped" }
        case "meeting-skip":
            guard run.status == "running" else { throw AppError("A turn must be running to skip it.") }
            run.skip = true
        default: throw AppError("Unknown Recall meeting action.")
        }
        return ["ok": true, "meeting": run.snapshot]
    }
}
